ujt/transformers.py

```diff
@@ -373,3 +373,44 @@ def apply_slis_to_node_map(sli_list, node_map):
         node.slis.extend(new_node_slis)
 
     return node_map
+
+
+def sort_nodes_by_parent_relationship(elements):
+    """Returns a list of elements where node parents always appear before node children.
+
+    For context, see https://github.com/plotly/dash-cytoscape/issues/112
+    and https://github.com/googleinterns/userjourneytool/issues/63
+
+    Args:
+        elements: a list of cytoscape elements.
+
+    Returns:
+        a list of topologically sorted cytoscape elements.
+    """
+    edges = []
+    node_id_element_map = {}
+    for element in elements:
+        if utils.is_node_element(element):
+            node_id_element_map[element["data"]["id"]] = element
+        else:
+            edges.append(element)
+
+    parent_child_map = defaultdict(list)
+    bfs_queue = deque()
+    # build a tree from parents to children
+    # (reversing the edge direction of cytoscape format)
+    for node in node_id_element_map.values():
+        node_id = node["data"]["id"]
+        if "parent" in node["data"]:
+            parent_id = node["data"]["parent"]
+            parent_child_map[parent_id].append(node_id)
+        else:
+            bfs_queue.append(node_id)
+
+    topologically_sorted_nodes = []
+    while bfs_queue:
+        node_id = bfs_queue.popleft()
+        bfs_queue.extend(parent_child_map[node_id])
```

Comment:
what are the chances of an infinite loop happening here? is it a problem if a client supplies a circular dependency list?

Reply:
The "edges" of this graph we're sorting are from the parent-child relationships of cytoscape graph elements (not the dependency relationship). I suppose we could have a loop, e.g. in protos received from the server, Node A declares its parent as Node B, Node B declares its parent as Node A. Then protos get converted to cytoscape elements with a loop in the parent-child relationship as well.

In other cases, I would lean away from validating the input here, since input with a loop violates the (admittedly implicit) input contract/interface of transformers, as the input list of isn't a valid set of cyto elements (not permissible by the library without this transformation). That is, I feel transformers of cyto elements should handle any valid set of renderable cyto elements and only validate additional constraints placed by the transformer itself. (e.g. if the UUID transformer relies on elements having a UUID appended to the element ids, it should validate that, because it's not a constraint imposed by the library, but rather the transformer itself)


However, since it could result in an infinite loop I think it's worth it to add some additional checks here. But I think the validation responsibility should be handled when the UJT first receives the topology from the server.